luarules/gadgets/game_disable_unit_sharing.lua

```diff
@@ -19,15 +19,41 @@ if not gadgetHandler:IsSyncedCode() then
 	return false
 end
 
-if not (Spring.GetModOptions().disable_unit_sharing
-			-- unit market handles the restriction instead if enabled so that selling still works
-			or Spring.GetModOptions().unit_market) then
+local unitSharingMode = Spring.GetModOptions().unit_sharing_mode or "enabled"
+local unitMarketEnabled = Spring.GetModOptions().unit_market
+
+-- If unit sharing is fully enabled and unit market isn't handling restrictions, disable this gadget
+if unitSharingMode == "enabled" and not unitMarketEnabled then
 	return false
 end
 
+local function isT2Constructor(unitDef)
+	if not unitDef then return false end
+
+	return not unitDef.isFactory
+			and #(unitDef.buildOptions or {}) > 0
+			and unitDef.customParams.techlevel == "2"
+end
+
 function gadget:AllowUnitTransfer(unitID, unitDefID, fromTeamID, toTeamID, capture)
-	if (capture) then
+	-- Always allow capture
+	if capture then
 		return true
 	end
-	return false
+
+	-- If unit market is handling this unit, allow the transfer
+	if unitMarketEnabled then
```

Comment:
As mentioned before, imo this gadget shouldn't be responsible for market concerns (**IF** the modoptions are orthogonal, one has priority over the other, otherwise if there's some intended mixing of behavior, e.g. allow t2 only on marketplace then sure, but imo this should be a setting on market - only allow t2 con market - instead of this gadgets)

Reply:
I agree. I will make this happen.

Reply:
Unfortunately because this is tied to an engine callback, and the way [gadgets](https://github.com/beyond-all-reason/Beyond-All-Reason/blob/2c69f12ff7a8353c16a6252da665be518887dce7/luarules/gadgets.lua#L1547) is written for this callback, make it somewhat difficult. That said it ALL points to this one gadget, so we probably could just rewrite this gadgets code.

Reply:
I understand the issue with gadget manager relaying true as default, and that using true|false as chain fallback information as faulty (perhaps gadgets that wish to not act upon this callback should return nil instead or some other value that relays the information correctly), but that's a problem to have when we have it.

I'm trying to understand the problem, what do you want to achieve and how so that you see this as a problem?